### PR TITLE
Character Movement!!!

### PR DESCRIPTION
**Pull request by prepsure**
Adds advanced character movement :)

## Main Additions
- Fully fledged character controller for basic movement, with options for custom middleware to control the movement and facing vectors
- An advanced movement state machine built on top of the character controller
- Controls and states for Walking, Jumping, Rolling, Long Jumping, Plunging, Gliding, and High Jumping
- Contextual acceleration and deceleration curves

## Testing
- Test in Scenes/Movement
- On keyboard, Shift is the action button, Space is jump, character movement is WASD, camera movement is the mouse
- On controller, the left and right face buttons are the action buttons, the bottom face button is jump, character movement is the left stick, camera movement is the right stick

# Extra Additions
- VectorMath library for performing operations on each component of a vector
- DictionaryUtil library for certain dictionary opertations
- Maid library for assisting with cleaning up events and running functions on destruction

### CODE REVIEW

File: Assets/Scripts/CharacterMechanics/CharacterMovement.cs [hunk 1]
```diff
@@ -0,0 +1,584 @@
...
+
+    // the position of the last platform on the previous frame
+    Vector3 LastPlatformLastPosition;
+
+    // the velocity of the last platform on the previous frame
+    Vector3 LastPlatformLastVelocity;
+    float LastPlatformLastTime;
+    #endregion
+
+    readonly Maid maid = new();
```
Comment (wormsbeeeee) [inline Assets/Scripts/CharacterMechanics/CharacterMovement.cs:204]: The Maid is really, really nice - I know it wasn't strictly necessary but I think that adding it makes the code much simpler and easier to deal with

File: Assets/Scripts/CharacterMechanics/CharacterMovement.cs [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+    // the last platform the character landed on
+    GameObject LastPlatform;
+
+    // the position of the last platform on the previous frame
+    Vector3 LastPlatformLastPosition;
+
+    // the velocity of the last platform on the previous frame
+    Vector3 LastPlatformLastVelocity;
+    float LastPlatformLastTime;
+    #endregion
```
Comment (wormsbeeeee) [inline Assets/Scripts/CharacterMechanics/CharacterMovement.cs:202]: Is there a reason for tracking the last platform? These fields aren't being used right now, do you plan on using them later on?

Comment (prepsure) [reply Assets/Scripts/CharacterMechanics/CharacterMovement.cs:202]: yeah sorry, that was for implementing platform momentum transfer, and that's not in this version

File: Assets/Scripts/CharacterMechanics/CharacterMovement.cs
Comment (wormsbeeeee) [inline Assets/Scripts/CharacterMechanics/CharacterMovement.cs:1]: The CharacterMovement class is extremely well-written and organized, and having all the options for parameterization is fantastic and will help a great deal when we refine the game's movement

File: Assets/Scripts/CharacterMechanics/AdvancedMovement.cs
Comment (wormsbeeeee) [inline Assets/Scripts/CharacterMechanics/AdvancedMovement.cs:1]: I like that movement options like these are kept separate ("Advanced" movement as opposed to just "Movement"), it's probably pretty self-explanatory to do it this way in the first place but still - very nice. Makes things easier to read through and maintain.

File: Assets/Scripts/CharacterMechanics/AdvancedMovementStates/Glide.cs
Comment (wormsbeeeee) [inline Assets/Scripts/CharacterMechanics/AdvancedMovementStates/Glide.cs:1]: Glide works pretty nicely when I tested

File: Assets/Scripts/Libs/DictionaryUtil.cs
Comment (wormsbeeeee) [inline Assets/Scripts/Libs/DictionaryUtil.cs:1]: This is nice, I wanna use this on my inventory stuff

**Comment by wormsbeeeee** [comment]
This is such a nice system honestly, I also love having all the animation curves for advanced movements. I know there's tons of advanced movements already but just wondering, do you think we'll add any other ones?

**Pull request merged**